### PR TITLE
planner, expression: take NullFlag into consideration when optimize the `int non-const` <cmp > `non-int const`

### LINKED ISSUE #16679: incorrectly fetch null data
## Description
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

TiDB

```
mysql> create table t0(c0 int);
Query OK, 0 rows affected (0.07 sec)

mysql>  insert into t0 values(null);
Query OK, 1 row affected (0.01 sec)

mysql>  select * from t0 where ((!(1.5071004017670217e-01=t0.c0))) IS NULL;
Empty set (0.00 sec)

mysql> select ((!(1.5071004017670217e-01=null))) IS NULL;
+--------------------------------------------+
| ((!(1.5071004017670217e-01=null))) IS NULL |
+--------------------------------------------+
|                                          1 |
+--------------------------------------------+
1 row in set (0.00 sec)

```

TiDB and MySQL 8.0.19

![image](https://user-images.githubusercontent.com/3251373/79867182-81c90700-8410-11ea-90ac-63f0c4fbbed3.png)

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                         |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-305-g94011e6064
Git Commit Hash: 94011e6064fffc7df1057e2a9b2b866f429e7494
Git Branch: HEAD
UTC Build Time: 2020-04-21 11:03:51
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```




## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 300

## Mentor
 * @XuHuaiyu

### PR DESCRIPTION
**Pull request by TszKitLo40**
### What problem does this PR solve?

Issue Number: close #16679  

Problem Summary:

### What is changed and how it works?

What's Changed:
Fix bug when fetching null data.
How it Works:
Check the `NotNullFlag` of the not-int constant.

### Check List 
Tests 
- Unit test
- Integration test
- Manual test (same steps in the issue)

### Release note 
- Fix incorrectly fetching null data

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
The test failed because the optimization rule is changed, some plan will change according to it.
You can run `./run-tests.sh -r all` in `cmd/explaintest` to run the test and record the result to get the correct test result.

File: planner/core/expression_rewriter_test.go [hunk 1]
```diff
@@ -260,3 +260,19 @@ func (s *testExpressionRewriterSuite) TestPatternLikeToExpression(c *C) {
 	tk.MustQuery("select 0 like '0';").Check(testkit.Rows("1"))
 	tk.MustQuery("select 0.00 like '0.00';").Check(testkit.Rows("1"))
 }
+
+func (s *testExpressionRewriterSuite) TestFetchNullData(c *C) {
```
Comment (mmyj) [inline planner/core/expression_rewriter_test.go:264]: TestFetchNullData -> TestIssue16679, and please add a comment `// TestIssue16679 contains tests for https://github.com/pingcap/tidb/issues/16679.`

**Comment by TszKitLo40** [comment]
> The test failed because the optimization rule is changed, some plan will change according to it.
> You can run `./run-tests.sh -r all` in `cmd/explaintest` to run the test and record the result to get the correct test result.

Thanks for your comments, I have fixed the tests.

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -698,7 +698,7 @@ func (s *testIntegrationSuite) TestPartitionPruningForInExpr(c *C) {
 
 	tk.MustExec("use test")
 	tk.MustExec("drop table if exists t")
-	tk.MustExec("create table t(a int(11), b int) partition by range (a) (partition p0 values less than (4), partition p1 values less than(10), partition p2 values less than maxvalue);")
+	tk.MustExec("create table t(a int(11) not null, b int) partition by range (a) (partition p0 values less than (4), partition p1 values less than(10), partition p2 values less than maxvalue);")
```
Comment (mmyj) [inline planner/core/integration_test.go:701]: should we add a new test for the column `a` without `not null`?

File: expression/builtin_compare_test.go [hunk 3]
```diff
@@ -67,6 +67,25 @@ func (s *testEvaluatorSuite) TestCompareFunctionWithRefine(c *C) {
 		{"123456789123456789123456789.12345 < a", "0"},
 		{"-123456789123456789123456789.12345 < a", "1"},
 		{"'aaaa'=a", "eq(0, a)"},
+		//{"1.5071004017670217e-01=a", "eq(1.5071004017670217e-01, a)"},
```
Comment (mmyj) [inline expression/builtin_compare_test.go:70]: why comment here?

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -1295,7 +1295,7 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
 	isExceptional, finalArg0, finalArg1 := false, args[0], args[1]
 	isPositiveInfinite, isNegativeInfinite := false, false
 	// int non-constant [cmp] non-int constant
-	if arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon {
+	if mysql.HasNotNullFlag(args[0].GetType().Flag) && arg0IsInt && !arg0IsCon && !arg1IsInt && arg1IsCon {
```
Comment (mmyj) [inline expression/builtin_compare.go:1298]: `mysql.HasNotNullFlag(args[0].GetType().Flag)` -> `mysql.HasNotNullFlag(arg0Type.Flag)`

File: expression/builtin_compare.go [hunk 5]
```diff
@@ -1314,7 +1314,7 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
 		}
 	}
 	// non-int constant [cmp] int non-constant
-	if arg1IsInt && !arg1IsCon && !arg0IsInt && arg0IsCon {
+	if mysql.HasNotNullFlag(args[1].GetType().Flag) && arg1IsInt && !arg1IsCon && !arg0IsInt && arg0IsCon {
```
Comment (mmyj) [inline expression/builtin_compare.go:1317]: ditto

**Review by mmyj — CHANGES_REQUESTED** [review]

File: planner/core/integration_test.go (hunk 2 above)
Comment (TszKitLo40) [reply planner/core/integration_test.go:702]: I have added a new test for it.

**Comment by mmyj** [comment]
please resolve the conflicting files.

**Comment by mmyj** [comment]
PTAL @XuHuaiyu

**Comment by mmyj** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@TszKitLo40 
I think we should not forbid `refineArgs` for all the scenarios, otherwise, it may cause bad performance for the queries on the partitioned table.
e.g.
``` sql
-- table t with range partition [p0 (1~10)], [p1 (11~20)]
select * from t where par_col in (9);
```
Before this commit, we only scan the `p0` partition. But after this commit, we'll trigger a full table scan, it's unacceptable.

A naive rule for whether we need to forbid `refineArgs` can be defined as the following:
1. If it's an expression in `SelectionExec`, and this comparison function is not wrapped with other expressions, we should not forbid the `refineArgs` even if the `NotNullFlag` is not set.
2. For others, we forbid `refineArgs` when `NotNullFlag` is not set.

**Comment by xuyifangreeneyes** [comment]
@TszKitLo40 is there any update?

**Comment by TszKitLo40** [comment]
> @TszKitLo40 is there any update?
No, I have no idea how to know whether the value is `null` in `planner` phase.

**Comment by xuyifangreeneyes** [comment]
> > @TszKitLo40 is there any update?
> > No, I have no idea how to know whether the value is `null` in `planner` phase. @xuyifangreeneyes Can you give me some advice?

We cannot know whether the value is `null` in `planner` phase. But as @XuHuaiyu said, we can allow `refineArgs` in some certain scenarios to avoid bad performance. For example,
```
create table t0 (a int);
explain select * from t0 where a = 3.4;
```
Here is the result if we allow `refineArgs` for `a = 3.4`
```
tidb> explain select * from t0 where a = 3.4;
+-------------+---------+------+---------------+---------------+
| id          | estRows | task | access object | operator info |
+-------------+---------+------+---------------+---------------+
| TableDual_6 | 0.00    | root |               | rows:0        |
+-------------+---------+------+---------------+---------------+
1 row in set (0.00 sec)
```
And here is the result if we forbid it.
```
tidb> explain select * from t0 where a = 3.4;
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 8000.00  | root      |               | data:Selection_6               |
| └─Selection_6           | 8000.00  | cop[tikv] |               | eq(cast(test.t0.a), 3.4)       |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)
```
Notice that the where condition `a = 3.4` is either `false` or `null`, so no row would be selected.

Hence, one improvement we can achieve is that when the comparison function is the where condition in select statement, even if `NotNullFlag` is not set, we try `refineArgs`. If the result of `refineArgs` is constantly false, `refineArgs` can be allowed since the comparison function is null if the column value is null and the where condition won't be satisfied no matter it is evaluated as false or null.

**Comment by TszKitLo40** [comment]
> > > @TszKitLo40 is there any update?
> > > No, I have no idea how to know whether the value is `null` in `planner` phase. @xuyifangreeneyes Can you give me some advice?
> 
> We cannot know whether the value is `null` in `planner` phase. But as @XuHuaiyu said, we can allow `refineArgs` in some certain scenarios to avoid bad performance. For example,
> 
> ```
> create table t0 (a int);
> explain select * from t0 where a = 3.4;
> ```
> 
> Here is the result if we allow `refineArgs` for `a = 3.4`
> 
> ```
> tidb> explain select * from t0 where a = 3.4;
> +-------------+---------+------+---------------+---------------+
> | id          | estRows | task | access object | operator info |
> +-------------+---------+------+---------------+---------------+
> | TableDual_6 | 0.00    | root |               | rows:0        |
> +-------------+---------+------+---------------+---------------+
> 1 row in set (0.00 sec)
> ```
> 
> And here is the result if we forbid it.
> 
> ```
> tidb> explain select * from t0 where a = 3.4;
> +-------------------------+----------+-----------+---------------+--------------------------------+
> | id                      | estRows  | task      | access object | operator info                  |
> +-------------------------+----------+-----------+---------------+--------------------------------+
> | TableReader_7           | 8000.00  | root      |               | data:Selection_6               |
> | └─Selection_6           | 8000.00  | cop[tikv] |               | eq(cast(test.t0.a), 3.4)       |
> |   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
> +-------------------------+----------+-----------+---------------+--------------------------------+
> 3 rows in set (0.00 sec)
> ```
> 
> Notice that the where condition `a = 3.4` is either `false` or `null`, so no row would be selected.
> 
> Hence, one improvement we can achieve is that when the comparison function is the where condition in select statement, even if `NotNullFlag` is not set, we try `refineArgs`. If the result of `refineArgs` is constantly false, `refineArgs` can be allowed since the comparison function is null if the column value is null and the where condition won't be satisfied no matter it is evaluated as false or null.

But it is hard to define whether `refineArgs` should be allowed.

**Pull request closed without merging**